types/object.go

```diff
@@ -0,0 +1,603 @@
+package types
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/NeowayLabs/abad/internal/utf16"
+)
+
+type (
+	// DataObject is a collection of named values.
+	DataObject struct {
+		// Class is the kind of object
+		class         string
+		notExtensible bool
+		props         map[string]*PropertyDescriptor
+	}
+
+	callable interface {
+		Call(this *DataObject, args []Value) Value
+	}
+)
+
+var (
+	// S is an alias to make utf-16 strings.
+	S            = utf16.S
+	valueAttr    = S("value")
+	writableAttr = S("writable")
+	getAttr      = S("get")
+	setAttr      = S("set")
+	enumAttr     = S("enumerable")
+	cfgAttr      = S("configurable")
+
+	protoAttr    = S("prototype")
+	toStringAttr = S("toString")
+	valueOfAttr  = S("valueOf")
+)
+
+// DefaultPrototypeDesc is a base prototype object that extends Null.
+// The root of the prototype-based type hierarchy.
+func DefaultPrototypeDesc() *PropertyDescriptor {
+	return NewDataPropDesc(Null, true, false, false)
+}
+
+// NewDataObject creates a new DataObject using proto as
+// prototype.
+func NewDataObject(proto Value) *DataObject {
+	obj := NewBaseDataObject()
+
+	// obj must extend proto
+	delete(obj.props, "prototype")
+
+	// error ignored because it does not fail if
+	// there's no previous properties.
+	ok, err := obj.DefineOwnPropertyP(protoAttr,
+		NewDataPropDesc(proto, false, true, false), true)
```

Comment:
This is the moment where keyword arguments in functions makes sense to me, I have no idea of what means these whole false/true parameters here and it would be great if this was clear without me having to jump around the code (the programming environ can help a lot on this cases, but this got me thinking).